src/lib/storage/table.cpp

```diff
@@ -81,23 +82,55 @@ const std::string& Table::column_name(const ColumnID column_id) const { return _
 
 const std::string& Table::column_type(const ColumnID column_id) const { return _col_types.at(column_id); }
 
-Chunk& Table::get_chunk(ChunkID chunk_id) { return *_chunks.at(chunk_id); }
+Chunk& Table::get_chunk(ChunkID chunk_id) {
+  std::lock_guard<std::mutex> lock(_chunk_lock);
+  return *_chunks.at(chunk_id);
+}
+
+const Chunk& Table::get_chunk(ChunkID chunk_id) const {
+  std::lock_guard<std::mutex> lock(_chunk_lock);
+  return std::as_const(_get_chunk(chunk_id));
+}
 
-const Chunk& Table::get_chunk(ChunkID chunk_id) const { return *_chunks.at(chunk_id); }
+Chunk& Table::_get_chunk(ChunkID chunk_id) const { return *_chunks.at(chunk_id); }
 
 void Table::print(std::ostream& out) const {
   int col_width = 20;
-  for (auto name : _col_names) {
+  for (const auto& name : _col_names) {
     std::string value = name;
     if (static_cast<int>(value.length()) > col_width) value.resize(col_width);
     out << value << std::string(col_width - value.length(), ' ');
   }
   out << "\n" << std::string(col_width * column_count(), '-') << "\n";
-  for (auto chunk : _chunks) {
+  for (const auto& chunk : _chunks) {
     chunk->print(col_width);
   }
 }
 
-void Table::compress_chunk(ChunkID chunk_id) { throw std::runtime_error("Implement Table::compress_chunk"); }
+void Table::compress_chunk(ChunkID chunk_id) {
+  std::lock_guard<std::mutex> lock(_chunk_lock);
+  auto& uncompressed_chunk = _get_chunk(chunk_id);
+  Assert(uncompressed_chunk.size() == target_chunk_size(),
+         "Attempt to compress chunk that is not yet completely filled.");
+
+  std::shared_ptr<Chunk> compressed_chunk = std::make_shared<Chunk>();
+
+  for (ColumnID column_id = ColumnID{0}; column_id < column_count(); ++column_id) {
```

Comment:
We need every single nano second :D
```suggestion
  auto column_count = column_count();
  for (ColumnID column_id = ColumnID{0}; column_id < column_count; ++column_id) {
```